src/plugins/kibana_react/public/code_editor/code_editor.tsx

```diff
@@ -154,14 +165,63 @@ export class CodeEditor extends React.Component<Props, {}> {
 
     this._editor = editor;
 
+    const domNode = editor.getDomNode();
+    if (domNode) {
+      domNode
+        .getElementsByTagName('textarea')[0]
+        .setAttribute('aria-describedby', this.ariaTabFocusId);
+    }
+
+    this._syncTabFocusMode();
+
+    editor.addCommand(
+      monaco.KeyCode.Escape,
+      () => {
+        editor.trigger('Change tab focus mode', 'editor.action.toggleTabFocusMode', {});
+
+        this._syncTabFocusMode();
+      },
+      '!suggestWidgetVisible'
+    );
+
     if (this.props.editorDidMount) {
       this.props.editorDidMount(editor);
     }
   };
 
+  _syncTabFocusMode = () => {
+    if (!this._editor) {
+      return;
+    }
+
+    const { tabFocusMode } = this._editor.getConfiguration();
+    const extraEditorClassName =
+      this.props.options && this.props.options.extraEditorClassName
+        ? this.props.options.extraEditorClassName
+        : '';
+
+    this.setState({
+      tabFocusMode,
+    });
+
+    this._editor.updateOptions({
+      extraEditorClassName: `${extraEditorClassName} ${
+        tabFocusMode ? 'monacoEditor--tabMode' : ''
+      }`,
+    });
+
+    this.setState({
+      tabFocusMode,
+    });
+  };
+
   render() {
     const { languageId, value, onChange, width, height, options } = this.props;
 
+    const tabFocusMessage = this.state.tabFocusMode
+      ? 'Tab moves focus. To toggle tab mode, press escape'
```

Comment:
I think a writer could also help us here, but I have a couple suggestions to improve this copy so it's easier for people to understand the change in behavior.

1. I suggest rephrasing the description of the behavior as an action the user can perform.
2. I also suggest using language that's symmetrical.

Here's an example of how I think these changes would look:

> Press tab to move focus. Press escape to change tab to insert the tab character.

> Press tab to insert the tab character. Press escape to change tab to move focus.